src/fromager/metrics.py

```diff
@@ -0,0 +1,74 @@
+import functools
+import logging
+import time
+import typing
+from collections import defaultdict
+
+from packaging.requirements import Requirement
+from packaging.version import Version
+
+_time_store: dict[str, dict[str, float]] = defaultdict(dict[str, float])
+_time_description_store: dict[str, str] = {}
+
+
+def timeit(description: str):
+    def timeit_decorator(func: typing.Callable) -> typing.Callable:
+        _time_description_store[func.__name__] = description
+
+        @functools.wraps(func)
+        def wrapper_timeit(
+            *, req: Requirement | None = None, **kwargs: typing.Any
+        ) -> typing.Any:
+            start = time.perf_counter()
+            ret = func(req=req, **kwargs)
+            end = time.perf_counter()
+            # get the logger for the module from which this function was called
+            logger = logging.getLogger(func.__module__)
+            version = (
+                kwargs.get("version")
+                or kwargs.get("dist_version")
+                or kwargs.get("resolved_version")
+            )
+            if not version:
+                version = _extract_version_from_return(ret)
+
+            runtime = end - start
+
+            if req:
+                logger.debug(f"{req.name}: took {runtime:.2f} seconds to {description}")
+            else:
+                logger.debug(f"took {runtime:.2f} seconds to {description}")
+
+            if req and version:
+                # store total time spent calling that function for a particular version of that req
+                _time_store[f"{req.name}=={version}"][func.__name__] = (
+                    _time_store[f"{req.name}=={version}"].get(func.__name__, 0)
+                    + runtime
+                )
+
+            return ret
+
+        return wrapper_timeit
+
+    return timeit_decorator
+
+
+def summarize(prefix: str) -> None:
+    logger = logging.getLogger(__name__)
+    for req in sorted(_time_store.keys()):
+        total_time = sum(_time_store[req].values())
+        log = f"{prefix} {req} took {total_time:.2f} seconds total"
+        for fn_name, time_taken in _time_store[req].items():
+            log += f", {time_taken:.2f} seconds to {_time_description_store[fn_name]}"
```

Comment:
Some of the longer builds take on the order of an hour, so it would be good to show the duration in a format that accounts for that. [datetime.timedelta](https://docs.python.org/3/library/datetime.html#timedelta-objects) does that automatically when converted to a string.